File: src/Explorer.Core/Services/KeyBindingResolver.cs
using NoctusExplorer.Core.Models;

namespace NoctusExplorer.Core.Services;

public sealed class KeyBindingResolver
{
    // chord → command id
    private readonly Dictionary<KeyChord, string> _chordToCommand = [];
    // command id → chord (reverse lookup)
    private readonly Dictionary<string, KeyChord> _commandToChord = new(StringComparer.OrdinalIgnoreCase);

    public void LoadBindings(Dictionary<string, KeyChord> bindings)
    {
        _chordToCommand.Clear();
        _commandToChord.Clear();
        foreach (var (commandId, chord) in bindings)
        {
            _chordToCommand[chord] = commandId;
            _commandToChord[commandId] = chord;
        }
    }

    public string? Resolve(KeyChord chord)
        => _chordToCommand.GetValueOrDefault(chord);

    public KeyChord? GetBinding(string commandId)
        => _commandToChord.GetValueOrDefault(commandId);

    public void SetBinding(string commandId, KeyChord chord)
    {
        // Remove old binding for this command if any
        if (_commandToChord.TryGetValue(commandId, out var oldChord))
            _chordToCommand.Remove(oldChord);

        // Remove old command for this chord if any
        if (_chordToCommand.TryGetValue(chord, out var oldCmd))
            _commandToChord.Remove(oldCmd);

        _chordToCommand[chord] = commandId;
        _commandToChord[commandId] = chord;
    }

    public void RemoveBinding(string commandId)
    {
        if (_commandToChord.TryGetValue(commandId, out var chord))
        {
            _chordToCommand.Remove(chord);
            _commandToChord.Remove(commandId);
        }
    }
}

File: macos/Sources/NoctusCore/Services/KeyBindingResolver.swift
import Foundation

public final class KeyBindingResolver {
    private var chordToCommand: [KeyChord: String] = [:]
    private var commandToChord: [String: KeyChord] = [:]

    public init() {}

    public func loadBindings(_ bindings: [String: KeyChord]) {
        chordToCommand.removeAll()
        commandToChord.removeAll()
        for (commandId, chord) in bindings {
            chordToCommand[chord] = commandId
            commandToChord[commandId] = chord
        }
    }

    public func resolve(_ chord: KeyChord) -> String? {
        chordToCommand[chord]
    }

    public func getBinding(_ commandId: String) -> KeyChord? {
        commandToChord[commandId]
    }

    public func setBinding(_ commandId: String, chord: KeyChord) {
        if let oldChord = commandToChord[commandId] {
            chordToCommand.removeValue(forKey: oldChord)
        }
        if let oldCmd = chordToCommand[chord] {
            commandToChord.removeValue(forKey: oldCmd)
        }
        chordToCommand[chord] = commandId
        commandToChord[commandId] = chord
    }

    public func removeBinding(_ commandId: String) {
        if let chord = commandToChord.removeValue(forKey: commandId) {
            chordToCommand.removeValue(forKey: chord)
        }
    }
}
